lib/fluent/plugin/prometheus_metrics.rb

```diff
@@ -0,0 +1,79 @@
+module Fluent::Plugin
+  class PromMetricsAggregator
+    def initialize
+      @known_metrics = []
+      @merged_lines = []
+    end
+
+    def get_metric_name(line)
+      metric = ''
+      tokens = line.split(' ')
+      if line[0] == '#'
+        if ['HELP', 'TYPE'].include?(tokens[1])
+          metric = tokens[2]
+        end
+      else
+        metric = tokens[0].split('{')[0]
+      end
+      metric
+    end
+
+    def is_meta_line(line)
+      tokens = line.split(' ')
+      line[0] == '#' && ['HELP', 'TYPE'].include?(tokens[1])
+    end
+
+    def add_metrics(metrics)
+      current_metric = ''
+      insert_offset = 0
+      lines = metrics.split("\n")
+      for line in lines
+        parsed_metric = get_metric_name(line)
+        if parsed_metric == current_metric
+          # Known metric, insert at current position
+          @merged_lines.insert(insert_offset, line)
+          insert_offset += 1
+        else
+          if parsed_metric == ''
+            # Unknown item, insert at current position
+            @merged_lines.insert(insert_offset, line)
+            insert_offset += 1
+          elsif @known_metrics.include?(parsed_metric)
+            if is_meta_line(line)
+              # Forget about this metric (prevents insertion of existing TYPE / HELP)
+              parsed_metric = ''
+            else
+              # Known metric, need to find insert location
```

Comment:
This may work well, but how about defining a dedicated data structure to keep lines for each metrics?
If we have PrometheusMetrics class that has the following values, help, type, merged_lines and then map metric name to PrometheusMetrics, it might be much easier to understand.

Reply:
I agree this solution isn't the most elegant one.
My idea was to keep maintenance to a minimum: should new metric type or meta declaration be added by future Prometheus release, this code will still be able to fill merged_lines. Also I didn't want to have extra code to deal with metrics type specifics (summaries have _sum & _count metrics, histogram have _bucket, _sum and _count suffixes).

If you think maintenance won't be an issue, I can work on the solution you propose.

Reply:
To be cleared, my intention is not meant to parse each line semantically. Just read lines and collect comments and metrics lines pairs for each metrics, then output them.

If we have these lines

```
# HELP http_request_duration_seconds The HTTP request latencies in seconds.
# TYPE http_request_duration_seconds histogram
http_request_duration_seconds_bucket{code="200",worker_id="0",method="GET",le="0.005"} 58
http_request_duration_seconds_bucket{code="200",worker_id="0",method="GET",le="0.01"} 58
```

The first 2 lines are comment and the latter 2 lines are metrics for `http_request_duration_seconds`.
So if another server has the comments and metrics for `http_request_duration_seconds`, it's easier to merge them by merging only metrics lines.

Reply:
Good idea! Done in ec77f15513759bbb0a8bdd6a1a6aeb985f0762fa